src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonPatternDetector.cs

```diff
@@ -0,0 +1,307 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.Json
+{
+    /// <summary>
+    /// Helper class to detect json in string tokens in a document efficiently.
+    /// </summary>
+    internal class JsonPatternDetector
+    {
+        private const string _jsonName = "json";
+        private const string _methodNameOfInterest = "Parse";
+        private static readonly HashSet<string> _typeNamesOfInterest = new HashSet<string>
+        {
+            "Newtonsoft.Json.Linq.JToken",
+            "Newtonsoft.Json.Linq.JObject",
+            "Newtonsoft.Json.Linq.JArray"
+        };
+
+        private static readonly ConditionalWeakTable<SemanticModel, JsonPatternDetector> _modelToDetector =
+            new ConditionalWeakTable<SemanticModel, JsonPatternDetector>();
+
+        private readonly SemanticModel _semanticModel;
+        private readonly ISyntaxFactsService _syntaxFacts;
+        private readonly ISemanticFactsService _semanticFacts;
+        private readonly IVirtualCharService _virtualCharService;
+        private readonly ISet<INamedTypeSymbol> _typesOfInterest;
+
+        /// <summary>
+        /// Helps match patterns of the form: language=json
+        /// 
+        /// All matching is case insensitive, with spaces allowed between the punctuation.
+        /// </summary>
+        private static readonly Regex s_languageCommentDetector =
+            new Regex(@"lang(uage)?\s*=\s*json((\s*,\s*)(?<option>[a-zA-Z]+))*",
+                RegexOptions.ExplicitCapture | RegexOptions.IgnoreCase | RegexOptions.Compiled);
+
+        public JsonPatternDetector(
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFacts,
+            ISemanticFactsService semanticFacts,
+            IVirtualCharService virtualCharService,
+            ISet<INamedTypeSymbol> typesOfInterest)
+        {
+            _semanticModel = semanticModel;
+            _syntaxFacts = syntaxFacts;
+            _semanticFacts = semanticFacts;
+            _virtualCharService = virtualCharService;
+            _typesOfInterest = typesOfInterest;
+        }
+
+        public static JsonPatternDetector GetOrCreate(
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFacts,
+            ISemanticFactsService semanticFacts,
+            IVirtualCharService virtualCharService)
+        {
+            // Do a quick non-allocating check first.
```

Comment:
Once the type is trimmed, would it be better to make it a struct and simply create a new struct whenever needed (without a cache)?

Reply:
Yup.  Thta's why i went with thsi pattern.